src/function/gds/all_shortest_paths.cpp

```diff
@@ -0,0 +1,724 @@
+        std::unique_ptr<EdgeCompute> edgeCompute;
+        switch (outputType) {
+        case RJOutputType::DESTINATION_NODES:
+        case RJOutputType::LENGTHS: {
+            edgeCompute = std::make_unique<AllSPLengthsEdgeCompute>(pathLengthsFrontiers.get(),
+                spOutputs->ptrCast<AllSPMultiplicitiesOutputs>()->multiplicities.get());
+            break;
+        }
+        case RJOutputType::PATHS: {
+            edgeCompute = std::make_unique<AllSPPathsEdgeCompute>(pathLengthsFrontiers.get(),
+                spOutputs->ptrCast<VarLenOrAllSPPathsOutputs>()->parentPtrs.get());
+            break;
+        }
+        default:
+            throw RuntimeException("Unrecognized RJOutputType in "
+                                   "AllSPAlgorithm::getRJCompState() setting edgeCompute: " +
+                                   std::to_string(static_cast<uint8_t>(outputType)) + ".");
+        }
+        return std::make_unique<RJCompState>(std::move(pathLengthsFrontiers),
+            std::move(edgeCompute), std::move(spOutputs), std::move(outputWriter));
+    }
+};
+
+struct VarLenJoinsEdgeCompute : public EdgeCompute {
+    DoublePathLengthsFrontiers* doublePathLengthsFrontiers;
+    ParentPtrsAtomics* parentPtrs;
+    ParentPtrsBlock* parentPtrsBlock;
+    explicit VarLenJoinsEdgeCompute(DoublePathLengthsFrontiers* doublePathLengthsFrontiers,
+        ParentPtrsAtomics* parentPtrs)
+        : doublePathLengthsFrontiers{doublePathLengthsFrontiers}, parentPtrs{parentPtrs} {
+        parentPtrsBlock = parentPtrs->addNewBlock();
+    };
+
+    bool edgeCompute(nodeID_t curNodeID, nodeID_t nbrID) override {
+        // We should always update the nbrID in variable length joins
+        if (!parentPtrsBlock->hasSpace()) {
+            parentPtrsBlock = parentPtrs->addNewBlock();
+        }
+        parentPtrs->addParent(doublePathLengthsFrontiers->curIter.load(std::memory_order_relaxed),
+            parentPtrsBlock, nbrID /* child */, curNodeID /* parent */);
+        return true;
+    }
+
+    std::unique_ptr<EdgeCompute> clone() override {
+        return std::make_unique<VarLenJoinsEdgeCompute>(this->doublePathLengthsFrontiers,
+            this->parentPtrs);
+    }
+};
+
+/**
+ * Algorithm for parallel all shortest paths computation, so all shortest paths from a source to
+ * is returned for each destination. If paths are not returned, multiplicities indicate the number
+ * of paths to each destination.
+ */
+class VarLenJoinsAlgorithm final : public RJAlgorithm {
+
+public:
+    explicit VarLenJoinsAlgorithm()
+        : RJAlgorithm(RJOutputType::PATHS, RJInputType::HAS_LOWER_BOUND) {}
+    VarLenJoinsAlgorithm(const VarLenJoinsAlgorithm& other) : RJAlgorithm(other) {}
+
+    std::unique_ptr<GDSAlgorithm> copy() const override {
+        return std::make_unique<VarLenJoinsAlgorithm>(*this);
+    }
+
+    std::unique_ptr<RJCompState> getRJCompState(processor::ExecutionContext* executionContext,
```

Comment:
I don't understand why inheritance is connected to returning an object vs pointer but made the change. You can comment more here to clarify.

Reply:
So we always object than pointer. Pointer is needed in the following two cases

-  a field is optional. So u need nullptr.
- you want to hide the actual implementation, e.g. we always can `func.exec()` instead of `additionFunc.exec()`. In such case, it must be `Function*` because `Function` cannot be instantiate as an object.